Align to NIST PQC Draft Standards FIPS 203, 204, 205 (#60)
openpgp-pqc/draft-openpgp-pqc: draft-wussler-openpgp-pqc.md

National Institute of Standards and Technology (NIST) in mid 2022
{{NISTIR-8413}} as the result of the NIST Post-Quantum Cryptography
Standardization process initiated in 2016 {{NIST-PQC}}. Namely, these are
ML-KEM (formerly CRYSTALS-Kyber) as a Key Encapsulation Mechanism (KEM), a KEM
being a modern building block for public-key encryption, and ML-DSA (formerly
CRYSTALS-Dilithium) as well as SLH-DSA (formerly SPHINCS+) as signature
schemes.

For the two ML-* schemes, this document follows the conservative strategy to
deploy post-quantum in combination with traditional schemes such that the
security is retained even if all schemes but one in the combination are broken.

> In contrast, the hashed-based signature scheme SLH-DSA is considered to be

Author: Good catch

be sufficiently well understood with respect to its security assumptions in
order to be used standalone. To this end, this document specifies the following

> new set: SLH-DSA standalone and ML-* as composite with ECC-based KEM and

Reviewer: ```suggestion
new set: SLH-DSA standalone and two ML-* as composite with ECC-based KEM and
```

Using regular expression like syntax in normal text is not necessarily common or intuitive to readers. Adding the word "two" makes it at least clearer that this is refering to two schemes and not denoting a specific one.

This section describes the individual post-quantum cryptographic schemes. All
schemes listed here are believed to provide security in the presence of a
cryptographically relevant quantum computer. However, the mathematical problems
on which the two CRYSTALS-* schemes and SPHINCS+ are based, are fundamentally
on which the two ML-* schemes and SLH-DSA are based, are fundamentally
different, and accordingly the level of trust commonly placed in them as well
as their performance characteristics vary.

> \[Note to the reader: This specification refers to the latest NIST submission

Reviewer: I wouldn't completely remove this editorial note yet, but only adapt it to the new situation that now we have NIST drafts for the algorithms. But it is still relevant to inform readers that these are only drafts and still subject to change.

Author: Team decision: Keep the note but switch from "NIST submission" to "NIST draft standards".

Todo @falko-strenzke

Reviewer: This is my proposal for the updated note:
``` 
\[Note to the reader: This specification refers to the NIST PQC standards drafts
Draft FIPS 203, Draft FIPS 204, and Draft FIPS 205 as if it were a final
specification. This is a temporary solution until the final versions of these
documents are available. The goal is to provide a sufficiently precise
specification of the algorithms already at the draft stage of this
specification, so that it is possible for implementers to create interoperable
implementations. Furthermore, we want to point out that, depending on possible
future changes to the draft standards by NIST, this specification may be updated
as soon as corresponding information becomes available.\]
```
Maybe we should indicate that we are referring to NIST **Drafts** also in the references by prefixing of postfixing the title either with "FIPS 203 (Initial Public Draft)" etc. as it is on the NIST web page of "FIPS 203 (Draft)" as it is in the PDF of the draft.

@fluppe2 Do you want to include the new version of the note into the PR (possibly after revising it if necessary from your point of view)? This is maybe easier than creating a separate PR or me working in this PR.

Author: Done in https://github.com/openpgp-pqc/draft-openpgp-pqc/pull/60/commits/6193c31b7ac32c11b9183d77ce7721f15aeae478

ML-KEM [FIPS-203] is based on the hardness of solving the learning-with-errors
problem in module lattices (MLWE). The scheme is believed to provide security
against cryptanalytic attacks by classical as well as quantum computers. This
specification defines ML-KEM only in composite combination with ECC-based
encryption schemes in order to provide a pre-quantum security fallback.

### ML-DSA {#mldsa-intro}

ML-DSA [FIPS-204] is a signature scheme that, like ML-KEM, is based on the

> hardness of solving lattice problems in module lattices. Accordingly, this

Author: I think the reason why we phrased it differently is because the security of ML-DSA bases on different problems (MLWE and MSIS). See §3.2 of FIPS-204. I guess there is space for optimization.

> hardness of solving lattice problems in module lattices. Accordingly, this

Author: @sehlen Are you fine with this formulation? FIPS-204 states that security goes back to a non-standard version of MSIS.

defined in the following subsections. Specifically, each of these subsections
defines the instances of the following operations:

    (eccCipherText, eccKeyShare) <- ECC-KEM.Encaps(eccPublicKey)

and

    (eccKeyShare) <- ECC-KEM.Decaps(eccPrivateKey, eccCipherText)

> To instantiate ECC-KEM, one must select a parameter set from

Reviewer 2: ```suggestion
To instantiate `ECC-KEM`, one must select a parameter set from
```

Author: Applied inline code marking also to ML-KEM in the anlagous sentence. See https://github.com/openpgp-pqc/draft-openpgp-pqc/pull/60/commits/fa1b1284d6bea9ceb1dafbbf470305744c0f0fe7

    chosen according to {{tab-ecdh-nist-artifacts}} or
    {{tab-ecdh-brainpool-artifacts}}

### ML-KEM {#mlkem-ops}

ML-KEM features the following operations:

    (mlkemCipherText, mlkemKeyShare) <- ML-KEM.Encaps(mlkemEncapsKey)

and

    (mlkemKeyShare) <- ML-KEM.Decaps(mlkemCipherText, mlkemDecapsKey)

The above are the operations `ML-KEM.Encaps` and `ML-KEM.Decaps` defined in
[FIPS-203]. Note that `mlkemEncapsKey` is public and `mlkemDecapsKey` is
private keying material.

ML-KEM has the parameterization with the corresponding artifact lengths in
octets as given in {{tab-mlkem-artifacts}}. All artifacts are encoded as
defined in [FIPS-203].

{: title="ML-KEM parameters artifact lengths in octets" #tab-mlkem-artifacts}
Algorithm ID reference | ML-KEM       | Public key | Secret key | Ciphertext | Key share
----------------------:| ------------ | ---------- | ---------- | ---------- | ---------
29, 31, 33             | ML-KEM-768   | 1184       | 2400       | 1088       | 32

> 30, 32, 34             | ML-KEM-1024  | 1568       | 3168       | 1568       | 32

Author: See [f5aeef2](https://github.com/openpgp-pqc/draft-openpgp-pqc/pull/60/commits/f5aeef293a70dd6cae838925b71e82e4f6e5e5b9).

> 30, 32, 34             | ML-KEM-1024  | 1568       | 3168       | 1568       | 32

Reviewer 2: Hmm, I see, but this breaks the plaintext version
![image](https://github.com/openpgp-pqc/draft-openpgp-pqc/assets/9117367/c4dae787-eff3-4891-b86d-567370901033)

> Algorithm ID reference | ML-KEM                    | Public key | Secret key | Ciphertext | Key share
> ----------------------:| ------------------------- | ---------- | ---------- | ---------- | ---------
> 29, 31, 33             | ML&#x2011;KEM&#x2011;768  | 1184       | 2400       | 1088       | 32
> 30, 32, 34             | ML&#x2011;KEM&#x2011;1024 | 1568       | 3168       | 1568       | 32

Reviewer 2: ```suggestion
Algorithm ID reference | ML-KEM      | Public key | Secret key | Ciphertext | Key share
----------------------:| ----------- | ---------- | ---------- | ---------- | ---------
29, 31, 33             | ML-KEM-768  | 1184       | 2400       | 1088       | 32
30, 32, 34             | ML-KEM-1024 | 1568       | 3168       | 1568       | 32
```

Author: See https://github.com/openpgp-pqc/draft-openpgp-pqc/pull/60#discussion_r1311121166

Author: Let it break as it break, no special character encodings.

To instantiate ML-KEM, one must select a parameter set from the column
"ML-KEM" of {{tab-mlkem-artifacts}}.

The procedure to perform `ML-KEM.Encaps()` is as follows:

 1. Extract the encapsulation key `mlkemEncapsKey` that is part of the
    recipient's composite public key

 2. Invoke `(mlkemCipherText, mlkemKeyShare) <- ML-KEM.Encaps(mlkemEncapsKey)`

 3. Set `mlkemCipherText` as the ML-KEM ciphertext

 4. Set `mlkemKeyShare` as the ML-KEM symmetric key share

The procedure to perform `ML-KEM.Decaps()` is as follows:

 1. Invoke `mlkemKeyShare <-  ML-KEM.Decaps(mlkemCipherText, mlkemDecapsKey)`

 2. Set `mlkemKeyShare` as the ML-KEM symmetric key share

## Composite Encryption Schemes with ML-KEM {#ecc-mlkem}

{{kem-alg-specs}} specifies the following ML-KEM + ECC composite public-key
encryption schemes:

{: title="ML-KEM + ECC composite schemes" #tab-mlkem-ecc-composite}
Algorithm ID reference | ML-KEM       | ECC-KEM   | ECC-KEM curve
----------------------:| ------------ | --------- | --------------
29                     | ML-KEM-768   | x25519Kem | Curve25519

> 30                     | ML-KEM-1024  | x448Kem   | Curve448

Reviewer 2: Why changing `X*` to `Curve*`?

Author: Because the column lists the curves associated to the algorithm. X* denotes algorithms/functions and Curve* denotes the curves. I felt it should be consistent.

    //   customizationString - the UTF-8 encoding of the string "KDF"

    eccKemData = eccKeyShare || eccCipherText

>     mlkemKemData = mlkemKeyShare || mlkemCipherText

Author: Good point, I opted for the first suggestion.

 3. Check that the own and the extracted algorithm ID match

>  4. Parse the `eccSecretKey` and `mlkemDecapsKey` from the algorithm specific

Author: Good catch. Changed `eccSecretKey` to `eccPrivateKey`

And yes, I adopted the terminology from ML-KEM specification. This is a remark I forgot to write in the PR description above. We should discuss that. I introduced this terminology in §5.1.2, saying that the encapsulation key is public and the decapsulation key is private keying material. I think since we have composite public and private keys which are composed of the ECC private/public and the ML-KEM Decapsulation/Encapsulation keys it does no harm to adopt the terminology of NIST.

>  4. Parse the `eccSecretKey` and `mlkemDecapsKey` from the algorithm specific

Reviewer: team agreed on this suggestion on 2023-09-11

>  4. Parse the `eccSecretKey` and `mlkemDecapsKey` from the algorithm specific

Author: Team agrees on:
- eccPublicKey and eccSecretKey
- mlkemPublicKey and mlkemSecretKey
- ecdsaPublicKey and ecdsaSecretKey
- eddsaPublicKey and eddsaSecretKey

Todo @fluppe2

## Packet specifications

### Public-Key Encrypted Session Key Packets (Tag 1) {#ecc-mlkem-pkesk}

The algorithm-specific fields consists of:

 - A fixed-length octet string representing an ECC ephemeral public key in the
   format associated with the curve as specified in {{ecc-kem}}.

 - A fixed-length octet string of the ML-KEM ciphertext, whose length depends
   on the algorithm ID as specified in {{tab-mlkem-artifacts}}.

 - The one-octet algorithm identifier, if it is passed (in the case of a v3
   PKESK packet).

 - A variable-length field containing the wrapped session key:

   - A one-octet size of the following field;

   - The wrapped session key represented as an octet string, i.e., the output
     of the encryption procedure described in {{ecc-mlkem-encryption}}.

Note that unlike most public-key algorithms, in the case of a v3 PKESK packet,
the symmetric algorithm identifier is not encrypted.  Instead, it is prepended
to the encrypted session key in plaintext.  In this case, the symmetric

> algorithm used MUST be AES-128, AES-192 or AES-256 (algorithm ID 7, 8 or 9).

Author: ```suggestion
algorithm used MUST be AES-128, AES-192 or AES-256.
```
Is it really necessary to list the algorithm IDs for AES from another spec?

Reviewer 2: I don't think it's necessary, but makes the dev life slightly easier IMO

Author: So keep it as it was (listing the IDs)?
It's ok for me.

Reviewer 2: Yes, as much as we list the IDs in the hash binding tables

Author: Ok, then lets drop this suggestion.

### Key Material Packets {#mlkem-ecc-key}

The algorithm-specific public key is this series of values:

 - A fixed-length octet string representing an EC point public key, in the
   point format associated with the curve specified in {{ecc-kem}}.

>  - A fixed-length octet string containing the ML-KEM encapsulation key, whose

Reviewer: I am not sure about the change to "encapsulation/decapsulation key" from "public/private key". Otherwise, in order to be consistent, we would have to eliminate all uses of "public/private key" and replace it with "encryption/decryption key" or "verification/signature key". Does anyone see a specific reason why KEM keys should be handled differently in this respect?

39                     | brainpoolP256r1 | 32         | 65         | 32         | 32                | 32
40                     | brainpoolP384r1 | 48         | 97         | 48         | 48                | 48

### ML-DSA signatures {#mldsa-signature}

For ML-DSA signature generation the default hedged version of `ML-DSA.Sign`
given in [FIPS-204] is used. That is, to sign with ML-DSA the following
operation is defined:

    (mldsaSignature) <- ML-DSA.Sign(mldsaPrivateKey, dataDigest)

For ML-DSA signature verification the algorithm ML-DSA.Verify given in
[FIPS-204] is used.  That is, to verify with ML-DSA the following operation is
defined:

    (verified) <- ML-DSA.Verify(mldsaPublicKey, dataDigest, mldsaSignature)

ML-DSA has the parameterization with the corresponding artifact lengths in
octets as given in {{tab-mldsa-artifacts}}. All artifacts are encoded as
defined in [FIPS-204].

{: title="ML-DSA parameters and artifact lengths in octets" #tab-mldsa-artifacts}
Algorithm ID reference | ML-DSA    | Public key | Secret key | Signature value
----------------------:| --------- | -----------| ---------- | ---------------

> 35, 37, 39             | ML-DSA-65 | 1952       | 4000       | 3293

Author: See [f5aeef2](https://github.com/openpgp-pqc/draft-openpgp-pqc/pull/60/commits/f5aeef293a70dd6cae838925b71e82e4f6e5e5b9).

39                     | brainpoolP256r1 | 32         | 65         | 32         | 32                | 32
40                     | brainpoolP384r1 | 48         | 97         | 48         | 48                | 48

### ML-DSA signatures {#mldsa-signature}

For ML-DSA signature generation the default hedged version of `ML-DSA.Sign`
given in [FIPS-204] is used. That is, to sign with ML-DSA the following
operation is defined:

    (mldsaSignature) <- ML-DSA.Sign(mldsaPrivateKey, dataDigest)

For ML-DSA signature verification the algorithm ML-DSA.Verify given in
[FIPS-204] is used.  That is, to verify with ML-DSA the following operation is
defined:

    (verified) <- ML-DSA.Verify(mldsaPublicKey, dataDigest, mldsaSignature)

ML-DSA has the parameterization with the corresponding artifact lengths in
octets as given in {{tab-mldsa-artifacts}}. All artifacts are encoded as
defined in [FIPS-204].

{: title="ML-DSA parameters and artifact lengths in octets" #tab-mldsa-artifacts}

> Algorithm ID reference | ML-DSA                  | Public key | Secret key | Signature value
> ----------------------:| ----------------------- | -----------| ---------- | ---------------
> 35, 37, 39             | ML&#x2011;DSA&#x2011;65 | 1952       | 4000       | 3293
> 36, 38, 40             | ML&#x2011;DSA&#x2011;87 | 2592       | 4864       | 4595

Reviewer 2: ```suggestion
Algorithm ID reference | ML-DSA    | Public key | Secret key | Signature value
----------------------:| --------- | -----------| ---------- | ---------------
35, 37, 39             | ML-DSA-65 | 1952       | 4000       | 3293
36, 38, 40             | ML-DSA-87 | 2592       | 4864       | 4595
```

Author: @TJ-91 suggested to insert those hyphen characters in https://github.com/openpgp-pqc/draft-openpgp-pqc/pull/60#discussion_r1307226240 in order to avoid unwanted line breaks. Is it not breaking weirdly on your side?

Author: I opted for Aron's suggestion to avoid any kind of special character encodings. I think that ends up messy. Let it break as it breaks.

## Composite Signature Schemes with ML-DSA {#ecc-mldsa}

### Signature data digest {#mldsa-sig-data-digest}

> Signature data is digested prior to signing operations, see

Author: ```suggestion
Signature data (i.e. the data to be signed) is digested prior to signing operations, see
```

   depends on the algorithm ID as specified in {{tab-mldsa-artifacts}}.

# SLH-DSA

## The SLH-DSA Algorithms {#slhdsa}

The following table describes the SLH-DSA parameters and artifact lengths:

{: title="SLH-DSA parameters and artifact lengths in octets. The values equally apply to the parameter IDs of SLH-DSA-SHA2 and SLH-DSA-SHAKE." #slhdsa-artifact-lengths}
Parameter ID reference | Parameter name suffix | SLH-DSA public key | SLH-DSA secret key | SLH-DSA signature
----------------------:| ---------------------:| ------------------ | ------------------ | ------------------
1                      | 128s                  | 32                 | 64                 | 7856
2                      | 128f                  | 32                 | 64                 | 17088
3                      | 192s                  | 48                 | 96                 | 16224
4                      | 192f                  | 48                 | 96                 | 35664
5                      | 256s                  | 64                 | 128                | 29792
6                      | 256f                  | 64                 | 128                | 49856

### Signature Data Digest

> SLH-DSA signatures MUST use the associated hash as specified in

Author: See [6095c0f](https://github.com/openpgp-pqc/draft-openpgp-pqc/pull/60/commits/6095c0f84f52a68f60f637c592545d909eccedf4).

{{tab-slhdsa-hash}} for the digest of signature data according to

> {{I-D.ietf-openpgp-crypto-refresh}} Section 5.2.4. Signatures using other hash

Author: I just turned the sentence around. I want to say that the signature data digest before the signature algorithm must be done via a hash algorithm from the table. Aron called this "binding hashes".

> {{I-D.ietf-openpgp-crypto-refresh}} Section 5.2.4. Signatures using other hash

Reviewer: Is "signature data" really the correct term for the data that is signed? To me it seems it could also be understood as the data that represents the signature. Maybe we should better call it "to-be-signed data"?

> {{I-D.ietf-openpgp-crypto-refresh}} Section 5.2.4. Signatures using other hash

Reviewer: As @fluppe2 pointed out, the above use of the term "signature data" is consistent with the [crypto refresh](https://www.ietf.org/archive/id/draft-ietf-openpgp-crypto-refresh-10.html#section-5.2.4).

> {{I-D.ietf-openpgp-crypto-refresh}} Section 5.2.4. Signatures using other hash

Reviewer: I suggest to add a clarifying subclause "for the digest of the signature data (data to be signed)"

# SLH-DSA

## The SLH-DSA Algorithms {#slhdsa}

The following table describes the SLH-DSA parameters and artifact lengths:

{: title="SLH-DSA parameters and artifact lengths in octets. The values equally apply to the parameter IDs of SLH-DSA-SHA2 and SLH-DSA-SHAKE." #slhdsa-artifact-lengths}
Parameter ID reference | Parameter name suffix | SLH-DSA public key | SLH-DSA secret key | SLH-DSA signature
----------------------:| ---------------------:| ------------------ | ------------------ | ------------------
1                      | 128s                  | 32                 | 64                 | 7856
2                      | 128f                  | 32                 | 64                 | 17088
3                      | 192s                  | 48                 | 96                 | 16224
4                      | 192f                  | 48                 | 96                 | 35664
5                      | 256s                  | 64                 | 128                | 29792
6                      | 256f                  | 64                 | 128                | 49856

### Signature Data Digest {#slhdsa-sig-data-digest}

> Signature data is digested prior to signing operations, see

Author: ```suggestion
Signature data (i.e. the data to be signed) is digested prior to signing operations, see
```
As discussed in team on 2023-09-11.

# SLH-DSA

## The SLH-DSA Algorithms {#slhdsa}

The following table describes the SLH-DSA parameters and artifact lengths:

{: title="SLH-DSA parameters and artifact lengths in octets. The values equally apply to the parameter IDs of SLH-DSA-SHA2 and SLH-DSA-SHAKE." #slhdsa-artifact-lengths}
Parameter ID reference | Parameter name suffix | SLH-DSA public key | SLH-DSA secret key | SLH-DSA signature
----------------------:| ---------------------:| ------------------ | ------------------ | ------------------
1                      | 128s                  | 32                 | 64                 | 7856
2                      | 128f                  | 32                 | 64                 | 17088
3                      | 192s                  | 48                 | 96                 | 16224
4                      | 192f                  | 48                 | 96                 | 35664
5                      | 256s                  | 64                 | 128                | 29792
6                      | 256f                  | 64                 | 128                | 49856

### Signature Data Digest {#slhdsa-sig-data-digest}

> Signature data (i.e. the data to be signed) is digested prior to signing operations, see
> {{I-D.ietf-openpgp-crypto-refresh}} Section 5.2.4. SLH-DSA signatures MUST use
> the associated hash algorithm as specified in {{tab-slhdsa-hash}} for the
> signature data digest. Signatures using other hash algorithms MUST be
> considered invalid.

Reviewer 2: ```suggestion
Signature data (i.e. the data to be signed) is digested prior to signing
operations, see {{I-D.ietf-openpgp-crypto-refresh}} Section 5.2.4. SLH-DSA
signatures MUST use the associated hash algorithm as specified in
{{tab-slhdsa-hash}} for the signature data digest. Signatures using other hash
algorithms MUST be considered invalid.
```

uncorrelated, this is negligible.

In consequence, the ciphertexts already work sufficiently well as
domain-separator.

## SLH-DSA Message Randomizer {#slhdsa-sec-cons}

The specification of SLH-DSA {{FIPS-205}} prescribes an optional
non-deterministic message randomizer. This is not used in this specification,
as OpenPGP v6 signatures already provide a salted signature data digest of the
appropriate size.

## Binding hashes in signatures with signature algorithms

In order not to extend the attack surface, we bind the hash algorithm used for
signature data digestion to the hash algorithm used internally by the signature
algorithm.

ML-DSA internally uses a SHAKE256 digest, therefore we require SHA3 in the
ML-DSA + ECC signature packet, see {{mldsa-sig-data-digest}}.

In the case of SLH-DSA the internal hash algorithm varies based on the
algorithm and parameter ID, see {{slhdsa-sig-data-digest}}.

# Additional considerations

## Performance Considerations for SLH-DSA {#performance-considerations}

This specification introduces both ML-DSA + ECC as well as SLH-DSA as PQ(/T)
signature schemes.

Generally, it can be said that ML-DSA + ECC provides a performance in terms of

> execution time and space requirements that is close to that of traditional ECC

Reviewer: Hmm, I think regarding execution time of ML-DSA the statement could be held up, since according to our comparison it faster then conventional curves but considerably slower than EdDSA.

But with respect to space requirements, I don't think the statement can be held: For example for 128-bit security, Dilithium signatures are about 37x larger than EC signatures, while SLH-DSA small parameters only bring another factor of 3. 

If we compare ML-DSA with RSA (3072 bit), then ML-DSA signatures are 6x larger.

```suggestion
execution time requirements that is close to that of traditional ECC signature schemes.
Regarding the size of signatures and public keys, though, ML-DSA has far greater requirements than traditional schemes like EC-based or even RSA
```

signature schemes. Implementers may want to offer SLH-DSA for applications
where a higher degree of trust in the signature scheme is required. However,
SLH-DSA has performance characteristics in terms of execution time of the

> signature generation as well as space requirements for the signature that can

Reviewer: ```suggestion
signature generation as well as space requirements for the signature that are
```

> be, depending on the parameter choice, far greater than those of traditional or

Reviewer: ```suggestion
even greater than those of
```
